### ISSUE TITLE
Improve torch operator coverage (as exported by torch-mlir)

### ISSUE THREAD

**Issue by j2kun** [google/heir#2923]
In a separate repository https://github.com/j2kun/heir_torch_coverage, I created a set of torch-mlir export scripts, one for each torch operator.

For Conv2d for example, the [operators/conv2d](https://github.com/j2kun/heir_torch_coverage/tree/main/operators/conv2d) contains a single-op torch model with garbage weights, exported to MLIR and annotated with secret before being run through `heir-opt` (2026-05-01 release). If it succeeds, the resulting openfhe MLIR is saved, otherwise an erorr log is saved.

I let AI decide on what "representative" set of torch operators should be included in the list (there are over 2k torch operators), so if new ones are needed I can add them.

So this issue exists to track progress on supporting these operators in HEIR. Specifically, if the relevant torch operator compiles to linalg that HEIR can handle _at all_ (even if inefficiently) then it is considered done for this issue.

I also created a new label [torch-support](https://github.com/google/heir/issues?q=is%3Aissue%20state%3Aopen%20label%3A%22torch%20support%22) for tickets related to this.

Comment https://github.com/google/heir/issues/2923#issuecomment-4391907478 has a regularly-updated table containing results of a CI that tests all the torch operators in the test suite.

You can find the failure logs by navigating to the appropriate GH action run of https://github.com/google/heir/actions/workflows/torch_mlir_coverage.yml and downloading the "failure-logs" artifact.

The torch-ci models tested are checked in at https://github.com/google/heir/tree/main/scripts/torch_ci

**Comment by j2kun** [comment]
| Operator | Status | Details |
| --- | --- | --- |
| add | PASS |  |
| avgpool1d | FAIL (heir-opt) | error: 'linalg.conv_2d_nchw_fchw' op Expected equal strides for Conv2DNchwFchwOp     %3 = linalg.poo |
| avgpool2d | PASS |  |
| avgpool3d | FAIL (heir-opt) | error: 'tensor_ext.assign_layout' op requires tensor rank to match the layout's domain size, but fou |
| batchnorm1d | PASS |  |
| batchnorm2d | PASS |  |
| batchnorm3d | PASS |  |
| bmm | FAIL (heir-opt) |  |
| cat | FAIL (heir-opt) | error: unexpected unrealized conversion cast op found     %concat = tensor.concat dim(1) %arg0, %arg |
| conv1d | PASS |  |
| conv2d | FAIL (heir-opt) | error: unexpected unrealized conversion cast op found   func.func @main(%arg0: tensor<1x3x16x16xf32> |
| conv3d | FAIL (heir-opt) | error: 'linalg.conv_3d_ncdhw_fcdhw' op expected operand #0 rank (2) to match the result rank of inde |
| div | FAIL (heir-opt) | error: 'secret.generic' op ciphertext division is not supported in CKKS; expected plaintext divisor  |
| elu | FAIL (heir-opt) | error: 'lattigo.ckks.encode' op operand #1 must be Tensor or MemRef of floating-point or complex-typ |
| eq | FAIL (heir-opt) | error: failed to legalize operation 'secret.generic' that was explicitly marked illegal   func.func  |
| exp | PASS |  |
| flatten | PASS |  |
| gelu | FAIL (heir-opt) | error: 'secret.generic' op ciphertext division is not supported in CKKS; expected plaintext divisor  |
| gt | FAIL (heir-opt) | error: 'lattigo.ckks.decode' op operand #2 must be Tensor or MemRef of floating-point or complex-typ |
| leakyrelu | FAIL (heir-opt) | error: 'lattigo.ckks.encode' op operand #1 must be Tensor or MemRef of floating-point or complex-typ |
| linear | PASS |  |
| log | PASS |  |
| lt | FAIL (heir-opt) | error: 'lattigo.ckks.decode' op operand #2 must be Tensor or MemRef of floating-point or complex-typ |
| matmul | FAIL (heir-opt) | error: Only secret vectors and plaintext matrices are supported for linalg.vecmat     %2 = linalg.ma |
| maxpool1d | FAIL (heir-opt) | error: 'tensor_ext.assign_layout' op requires tensor rank to match the layout's domain size, but fou |
| maxpool2d | FAIL (heir-opt) | error: 'tensor_ext.assign_layout' op requires tensor rank to match the layout's domain size, but fou |
| maxpool3d | FAIL (heir-opt) | error: 'tensor_ext.assign_layout' op requires tensor rank to match the layout's domain size, but fou |
| mean | FAIL (heir-opt) | error: 'secret.generic' op ciphertext division is not supported in CKKS; expected plaintext divisor  |
| mm | FAIL (heir-opt) | error: 'linalg.matmul' op expected type of operand #2 ('tensor<1x1024xf32>') to match type of corres |
| mul | PASS |  |
| neg | FAIL (heir-opt) | error: failed to legalize operation 'lwe.rnegate' that was explicitly marked illegal: %6 = "lwe.rneg |
| permute | PASS |  |
| prelu | FAIL (heir-opt) | error: 'lattigo.ckks.encode' op operand #1 must be Tensor or MemRef of floating-point or complex-typ |
| prod | PASS |  |
| relu | PASS |  |
| sigmoid | PASS |  |
| silu | PASS |  |
| softmax | FAIL (heir-opt) | error: 'tensor_ext.assign_layout' op requires tensor rank to match the layout's domain size, but fou |
| sqrt | PASS |  |
| squeeze | PASS |  |
| sub | PASS |  |
| sum | PASS |  |
| tanh | PASS |  |
| torch_chunk | PASS |  |
| torch_select | PASS |  |
| transpose | PASS |  |

**Comment by AlexanderViand** [comment]
I think it might be helpful to split these into layout/kernel heavy things (convolutions, transpose, flatten, etc) and approximation heavy things (softmax, GELU, etc)?

Also, let me create some (sub) issues for the specific ones we already had on our internal issue tracker so I can link them to a public issue.
